src/merkle/tiered_smt/mod.rs

```diff
@@ -189,6 +189,54 @@ impl TieredSmt {
         old_value
     }
 
+    // ITERATORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Returns an iterator over all inner nodes of this [TieredSmt] (i.e., nodes not at depths 16
+    /// 32, 48, or 64).
+    ///
+    /// The iterator order is unspecified.
+    pub fn inner_nodes(&self) -> impl Iterator<Item = InnerNodeInfo> + '_ {
+        self.nodes.iter().filter_map(|(index, node)| {
+            if is_inner_node(index) {
+                Some(InnerNodeInfo {
+                    value: node.into(),
+                    left: self.get_node_unchecked(&index.left_child()).into(),
+                    right: self.get_node_unchecked(&index.right_child()).into(),
+                })
+            } else {
+                None
+            }
+        })
+    }
+
+    /// Returns an iterator over upper leaves (i.e., depth = 16, 32, or 48) for this [TieredSmt].
+    ///
+    /// Each yielded item is a (node, key, value) tuple where key is a full un-truncated key (i.e.,
+    /// with key[3] element unmodified).
+    ///
+    /// The iterator order is unspecified.
+    pub fn upper_leaves(&self) -> impl Iterator<Item = (RpoDigest, RpoDigest, Word)> + '_ {
```

Comment:
The reason to differentiate between the two is that leaf hashes are computed differently between them. What we want to return from here is a leaf hash as well as its pre-image. And because structure of pre-images differs between upper and bottom leaves, we can't return reduce them tot the same form. 

> Had you considered flattening the `Vec<(RpoDigest, Word)>` for `bottom_leaves(..)` such that we can combine with `upper_leaves` to have a single api for returning a leaves iterator which returns an iterator with `Item = (RpoDigest, RpoDigest, Word)`.

In `Item = (RpoDigest, RpoDigest, Word)` what would be the first digest? If leaf hash, then this wouldn't make sense for bottom leaves as the remaining `(RpoDigest, Word)` may not match this hash.